### PR TITLE
Archiving all-files scheduler

### PR DESCRIPTION
**Pull request by JasonWeill**
WIP fix for #349.

Creates an `AllFilesArchivingExecutionManager` and `AllFilesArchivingScheduler` class, with the intention of archiving all output files and side effects on job runs.

You can use these classes by running:

```
jupyter lab \
  --SchedulerApp.scheduler_class=jupyter_scheduler.scheduler.AllFilesArchivingScheduler \
  --Scheduler.execution_manager_class=jupyter_scheduler.executors.AllFilesArchivingExecutionManager
```

After the user chooses to download output files from a job, the side effect files are unpacked to wherever the first output file is located.

In the screen shot below, after downloading the output of a job that ran `writefile.ipynb`, which produces the side effect file `pun.txt`, the side effect file is in the `jobs` subdirectory.

image

### CODE REVIEW

File: jupyter_scheduler/scheduler.py
Comment (3coins) [inline jupyter_scheduler/scheduler.py:741]: I don't think we need to do any different here for run now jobs.

File: jupyter_scheduler/executors.py
Comment (dlqqq) [inline jupyter_scheduler/executors.py:252]: There's no need to call `os.path.abspath()` here, since the staging paths are all absolute paths.

File: jupyter_scheduler/executors.py
Comment (dlqqq) [inline jupyter_scheduler/executors.py:235]: There is a lot of similar code to `ArchivingExecutionManager`. I found that this made the code much more difficult to read, as I had to switch between the two implementations to discern the differences in behavior. I would highly encourage `AllFilesArchivingExecutionManager` to inherit from `ArchivingExecutionManager` and use both parent private methods and `super().execute()` to inherit logic and minimize code duplication.

File: jupyter_scheduler/job_files_manager.py
Comment (dlqqq) [inline jupyter_scheduler/job_files_manager.py:105]: The TGZ contains a flattened archive of the entire job staging directory at `<staging-dir>/<job-id>`, which also includes the input and output files. I think there's a significant opportunity to simplify the implementation here. Rather than extracting the input and output files from the tar, and then extracting the rest of the files separately via `tar.extractall(members=...)`, why not just call `tar.extractall()` with `members` unspecified?

This strategy would allow you to delete the `exclude_output_filepaths()` method and the `for` block above calling `tar.extractfile()` on each file type listed in `output_formats + ["input"]`.

File: jupyter_scheduler/executors.py
Comment (dlqqq) [inline jupyter_scheduler/executors.py:283]: 1. We should decide on whether we're OK with flattening all the side effect files and not preserving their directory structure. This will be a backwards-incompatible change, so we will want alignment on this.

2. The current implementation doesn't clean up side effect files generated under `<staging-dir>/<job-id>`. If we're OK with flattening all the side effect files, you could instead run the notebook under a temporary dir `<staging-dir>/<job-id>/.run` and delete all those files after building the tarball:

```py
working_dir = os.path.dirname(self.staging_paths["input"])
run_dir = os.path.join(working_dir, ".run")

        try:
            ep.preprocess(nb, {"metadata": {"path": run_dir }})
        except CellExecutionError as e:
            pass
        finally:
            # ... shutil.rmtree(run_dir)
```

I would also rename `working_dir` to `local_staging_dir` for clarity.

File: jupyter_scheduler/scheduler.py
Comment (dlqqq) [inline jupyter_scheduler/scheduler.py:719]: This class appears to be logically identical to `ArchivingScheduler`, with two exceptions:

1. The default `execution_manager_class` is set to `AllFilesArchivingExecutionManager`

2. Minor readability changes right before the `return` statement.

I would instead apply the readability changes to `ArchivingScheduler`, have `AllFilesArchivingScheduler` inherit from that, and just override the trait's default by leaving the definition there. Then you can delete the duplicate `get_staging_paths()` implementation.

If I'm wrong about the two methods being logically identical, please see if you can use inheritance to minimize code duplication.

**Review by dlqqq — CHANGES_REQUESTED** [review]
@JasonWeill Thank you for working on this! Left some feedback.

File: jupyter_scheduler/job_files_manager.py
Comment (dlqqq) [reply jupyter_scheduler/job_files_manager.py:72]: This may not be necessary; see https://github.com/jupyter-server/jupyter-scheduler/pull/388#discussion_r1289281656

**Review by dlqqq — CHANGES_REQUESTED** [review]
There is also a failing unit test blocking CI:

```
FAILED jupyter_scheduler/tests/test_job_files_manager.py::test_downloader_download[output_formats1-output_filenames1-staging_paths1-/home/runner/work/jupyter-scheduler/jupyter-scheduler/jupyter_scheduler/tests/test_files_output-False] - AssertionError: assert False
 +  where False = <function exists at 0x7feb182f7420>('/home/runner/work/jupyter-scheduler/jupyter-scheduler/jupyter_scheduler/tests/test_files_output/helloworld-out.ipynb')
 +    where <function exists at 0x7feb182f7420> = <module 'posixpath' (frozen)>.exists
 +      where <module 'posixpath' (frozen)> = os.path
```

No need for an urgent fix here, since CI is failing on `main` anyways until we merge #417. However, this definitely should be fixed after rebasing and before merging.

File: jupyter_scheduler/job_files_manager.py
Comment (JasonWeill) [reply jupyter_scheduler/job_files_manager.py:72]: My choice of the `members` name comes from Python's `tarfile` examples: https://docs.python.org/3/library/tarfile.html#examples

I'm also OK with simplifying the logic if the more specific function I wrote is not necessary.

**Comment by JasonWeill** [comment]
Simplifies archiving and unarchiving logic per @dlqqq. Seeking external feedback about having these new classes replace the old `Archiving*` classes; if there's no opposition, I can work on this tomorrow. Also, we need to document these new classes and how to use them.

**Review by dlqqq — APPROVED** [review]
Awesome work! 🎉

**Pull request merged**
